src/main/python/mmlspark/cyber/ml/access_anomalies/collaborative_filtering.py

```diff
@@ -0,0 +1,721 @@
+    """
+    def __init__(self,
+                 tenant_col: str,
+                 user_col: str,
+                 user_vec_col: str,
+                 res_col: str,
+                 res_vec_col: str,
+                 user_model_df: DataFrame,
+                 res_model_df: DataFrame):
+
+        self.tenant_col = tenant_col
+        self.user_col = user_col
+        self.user_vec_col = user_vec_col
+        self.res_col = res_col
+        self.res_vec_col = res_vec_col
+        self.user_model_df = user_model_df
+        self.res_model_df = res_model_df
+
+    def replace_models(self, user_model_df: Optional[DataFrame] = None, res_model_df: Optional[DataFrame] = None):
+        """
+        create a new model replacing the user and resource models with new ones (optional)
+
+        :param user_model_df: optional new user model mapping names to latent vectors
+        :param res_model_df: optional new resource model mapping names to latent vectors
+        :return:
+        """
+        return UserResourceCfDataframeModel(
+            self.tenant_col,
+            self.user_col,
+            self.user_vec_col,
+            self.res_col,
+            self.res_vec_col,
+            user_model_df if user_model_df is not None else self.user_model_df,
+            res_model_df if res_model_df is not None else self.res_model_df
+        )
+
+    def check(self):
+        """
+        check the validity of the model
+        :return: boolean value where True indicating the verification succeeded
+        """
+        return self._check_user_model() and self._check_res_model()
+
+    def _check_user_model(self):
+        field_map = {ff.name: ff for ff in self.user_model_df.schema.fields}
+
+        assert field_map.get(self.tenant_col) is not None, field_map
+        assert field_map.get(self.user_col) is not None
+
+        return self.user_model_df.select(
+            self.tenant_col, self.user_col
+        ).distinct().count() == self.user_model_df.count()
+
+    def _check_res_model(self):
+        field_map = {ff.name: ff for ff in self.res_model_df.schema.fields}
+
+        assert field_map.get(self.tenant_col) is not None, field_map
+        assert field_map.get(self.res_col) is not None
+
+        return self.res_model_df.select(
+            self.tenant_col, self.res_col
+        ).distinct().count() == self.res_model_df.count()
+
+
+class AccessAnomalyModel(Transformer):
+    """
+    A pyspark.ml.Transformer model that can predict anomaly scores for user, resource access pairs
+    """
+    def __init__(self, user_res_cf_df_model: UserResourceCfDataframeModel, output_col: str):
+        self.user_res_cf_df_model = user_res_cf_df_model
+        self.output_col = output_col
+
+    @property
+    def tenant_col(self):
+        return self.user_res_cf_df_model.tenant_col
+
+    @property
+    def user_col(self):
+        return self.user_res_cf_df_model.user_col
+
+    @property
+    def user_vec_col(self):
```

Comment:
How do the user and user_vec columns differ? Same for res and res_vec

Reply:
'user' column refers to the name (or id) of the user and user_vec refers to the latent feature vector calculated by the model (for that user).
Same for 'res'